backends/dpdk/dpdkAsmOpt.cpp

```diff
@@ -264,5 +264,236 @@ bool ValidateTableKeys::preorder(const IR::DpdkAsmProgram *p) {
+}
+
+void CopyPropagationAndElimination::elimCastOrMov(const IR::DpdkAsmStatement* stmt,
+                                        IR::IndexedVector<IR::DpdkAsmStatement>& instr) {
+    const IR::Expression* srcExpr = nullptr, *dstExpr = nullptr;
+    if (stmt->is<IR::DpdkMovStatement>()) {
+        auto mv = stmt->to<IR::DpdkMovStatement>();
+        srcExpr = mv->src;
+        dstExpr = mv->dst;
+    } else {
+        auto mv = stmt->to<IR::DpdkCastStatement>();
+        srcExpr = mv->src;
+        dstExpr = mv->dst;
+    }
+    auto src = srcExpr->toString();
+    auto dst = dstExpr->toString();
+    bool dropCopy = false;
+    if (dontEliminate.count(dst) == 0
+            && dst.startsWith("m.")
+            && newUsesInfo[dst] == 0
+            && haveSingleUseDef(dst)) {
+        dropCopy = true;  // do nothing, drop the statement
+    }
+    if (!dropCopy) {
+        auto r = replaceIfCopy(srcExpr);
+        if (dst != r->toString() && dst.startsWith("m.")) {
+            newUsesInfo[r->toString()]++;
+            instr.push_back(new IR::DpdkMovStatement(dstExpr, r));
+        } else {
+            instr.push_back(stmt);
+            newUsesInfo[src]++;
+        }
+    }
+}
+
+void CopyPropagationAndElimination::collectUseDef(IR::IndexedVector<IR::DpdkAsmStatement> stmts) {
+    for (auto stmt : stmts) {
+        if (auto mv = stmt->to<IR::DpdkMovStatement>()) {
+            markUseDef(mv);
+        } else if (auto c = stmt->to<IR::DpdkCastStatement>()) {
+            markUseDef(c);
+        } else if (auto b = stmt->to<IR::DpdkBinaryStatement>()) {
+            markUseDef(b);
+        } else if (auto c = stmt->to<IR::DpdkUnaryStatement>()) {
+            markUseDef(c);
+        } else if (auto jc = stmt->to<IR::DpdkJmpCondStatement>()) {
+            markUseDef(jc);
+        } else if (auto l = stmt->to<IR::DpdkLearnStatement>()) {
+            markUseDef(l);
+        } else if (auto e = stmt->to<IR::DpdkEmitStatement>()) {
+            auto type = typeMap->getType(e->header)->to<IR::Type_Header>();
+            if (type)
+                for (auto f : type->fields) {
+                    cstring name = e->header->toString() + "." + f->name.toString();
+                    usesInfo[name]++;
+                }
+        } else if (auto e = stmt->to<IR::DpdkExtractStatement>()) {
+            auto type = typeMap->getType(e->header)->to<IR::Type_Header>();
+            if (type)
+                for (auto f : type->fields) {
+                    cstring name = e->header->toString() + "." + f->name.toString();
+                    defInfo[name]++;
+                }
+            if (e->length)
+                usesInfo[e->length->toString()]++;
+        } else if (auto l = stmt->to<IR::DpdkLookaheadStatement>()) {
+            auto type = typeMap->getType(l->header)->to<IR::Type_Header>();
+            if (type)
+                for (auto f : type->fields) {
+                    cstring name = l->header->toString() + "." +f->name.toString();
```

Comment:
so sometimes you use header fields, sometimes full headers?

Reply:
have headers with structs been flattened?

Reply:
yes

Reply:
> so sometimes you use header fields, sometimes full headers?

It's not like that, because there won't be such uses.
Here I am trying to construct cstring for member expression combining header  name and it's fields.

i.e. consider below dpdk statement 
`emit h.ethernet`
above statement basically defines the all fields of ethernet header in the form like
`
h.ethernet.srcAddr
h.ethernet.dstAddr
...
`

Reply:
same goes for lookahead.